jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java

```diff
@@ -140,40 +143,43 @@ public ImagesAndRegistryClient call()
             LogEvent.lifecycle(
                 "The base image requires auth. Trying again for " + imageReference + "..."));
 
-        Credential registryCredential =
+        Credential credential =
             RegistryCredentialRetriever.getBaseImageCredential(buildContext).orElse(null);
-
         RegistryClient registryClient =
             buildContext
                 .newBaseImageRegistryClientFactory()
-                .setCredential(registryCredential)
+                .setCredential(credential)
                 .newRegistryClient();
 
-        try {
-          // TODO: refactor the code (https://github.com/GoogleContainerTools/jib/pull/2202)
-          if (registryCredential == null || registryCredential.isOAuth2RefreshToken()) {
-            throw ex;
-          }
-
-          eventHandlers.dispatch(LogEvent.debug("Trying basic auth for " + imageReference + "..."));
-          registryClient.configureBasicAuth();
+        String wwwAuthenticate = ex.getHttpResponseException().getHeaders().getAuthenticate();
+        if (wwwAuthenticate != null) {
+          eventHandlers.dispatch(
+              LogEvent.debug("WWW-Authenticate for " + imageReference + ": " + wwwAuthenticate));
+          registryClient.authPullByWwwAuthenticate(wwwAuthenticate);
           return new ImagesAndRegistryClient(
               pullBaseImages(registryClient, progressEventDispatcher), registryClient);
 
-        } catch (RegistryUnauthorizedException registryUnauthorizedException) {
-          // The registry requires us to authenticate using the Docker Token Authentication.
-          // See https://docs.docker.com/registry/spec/auth/token
-          eventHandlers.dispatch(
-              LogEvent.debug("Trying bearer auth for " + imageReference + "..."));
-          if (registryClient.doPullBearerAuth()) {
-            return new ImagesAndRegistryClient(
-                pullBaseImages(registryClient, progressEventDispatcher), registryClient);
+        } else {
+          // Not getting WWW-Authenticate is unexpected in practice, and we may just blame the
+          // server and fail. However, to keep some old behavior, try a few things as a last resort.
+          // TODO: consider removing this fallback branch.
+          if (credential != null && !credential.isOAuth2RefreshToken()) {
+            eventHandlers.dispatch(
```

Comment:
should we actually do `LogEvent.warn` here with some guidance on reporting this, so that we may identify misbehaving servers by user reports?

Reply:
I still think this is a good idea to identify registries that are acting strange. But up to you.

Reply:
Sorry, I missed this comment. I thought about that, and I agree logging a warning will be very valuable to us. But I think a warning should be something actionable and fixable on the user side, so that's my hesitation. I've seen some people regard a warning seriously (probably because of a 0-warning policy auto-enforced by some system). Gathering information is really useful to us, but I feel we should be careful for dual-purposing a warning.